accumulation of any changes in the xbrl validation tests

cmgosnell opened the pull request:
## PR Checklist

Before requesting a review of your pull request, please make sure you've done the
following:

* [x] Merge the most recent version of `dev` (or the appropriate upstream branch) into
      your branch and resolved any merge conflicts. You may need to do this several
      times over the course of a PR as `dev` changes frequently.
* [ ] Verify that all of the CI checks on your PR are passing. See
      [Running Tests with Tox](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
      for details on how to run the full test suite locally if you need to debug a
      particular failure.
* [x] Ensure that the docstrings for any new modules, classes, functions, or methods are
      descriptive enough for developers and users to understand your code.
* [x] If you expanded data coverage or changed the outputs, ensure that the full
      [data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
      pass locally on a fresh DB.
* [ ] If you've added new functions or classes, ensure that they have at least basic
      unit tests.
* [x] If you've added new analyses, make sure they include defensive sanity checks that
      will catch unexpected data issues.
* [ ] Update the
      [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html)
      to reflect your changes. Make sure to reference the PR and any related issues.
* [x] Do your own review of the PR. Add comments highlighting areas where you have
      questions you'd like reviewers to answer, known issues, solutions you're
      unsatisfied with, or other things that deserve special attention from the
      reviewer.

cmgosnell commented:
Current error from `tox -e nuke`

```
test/integration/datasette_metadata_test.py:20:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'pudl.metadata.classes.DatasetteMetadata'>, data_source_ids = ['pudl', 'ferc1', 'eia860', 'eia860m', 'eia923']
xbrl_ids = ['ferc1_xbrl', 'ferc2_xbrl', 'ferc6_xbrl', 'ferc60_xbrl', 'ferc714_xbrl'], extra_etl_groups = ['entity_eia', 'glue', 'static_eia', 'static_ferc1']
pudl_settings = {'censusdp1tract_db': 'sqlite:////Users/christinagosnell/code/pudl_work/sqlite/censusdp1tract.sqlite', 'data_dir': '/U.../v3s4wmyd75v5gjb636r5bvl80000gq/T/pytest-of-christinagosnell/pytest-36/pudl0/sqlite/ferc1_xbrl_datapackage.json'), ...}

    @classmethod
    def from_data_source_ids(
        cls,
        data_source_ids: Iterable[str] = [
            "pudl",
            "ferc1",
            "eia860",
            "eia860m",
            "eia923",
        ],
        xbrl_ids: Iterable[str] = [
            "ferc1_xbrl",
            "ferc2_xbrl",
            "ferc6_xbrl",
            "ferc60_xbrl",
            "ferc714_xbrl",
        ],
        extra_etl_groups: Iterable[str] = [
            "entity_eia",
            "glue",
            "static_eia",
            "static_ferc1",
        ],
        pudl_settings: dict = {},
    ) -> "DatasetteMetadata":
        """Construct a dictionary of DataSources from data source names.

        Create dictionary of first and last year or year-month for each source.

        Args:
            data_source_ids: ids of data sources currently included in Datasette
            xbrl_ids: ids of data converted XBRL data to be included in Datasette
            extra_etl_groups: ETL groups with resources that should be included
            pudl_settings: Dictionary of settings.
        """
        # Compile a list of DataSource objects for use in the template
        data_sources = [DataSource.from_id(ds_id) for ds_id in data_source_ids]

        # Instantiate all possible resources in a Package:
        pkg = Package.from_resource_ids()
        # Grab a list of just the resources we want to output:
        resources = [
            res
            for res in pkg.resources
            if res.etl_group in data_source_ids + extra_etl_groups
        ]

        # Get XBRL based resources
        for xbrl_id in xbrl_ids:
            # Read JSON Package descriptor from file
>           with open(pudl_settings[f"{xbrl_id}_datapackage"]) as f:
E           FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/p8/v3s4wmyd75v5gjb636r5bvl80000gq/T/pytest-of-christinagosnell/pytest-36/pudl0/sqlite/ferc1_xbrl_datapackage.json'

.env_tox/lib/python3.10/site-packages/pudl/metadata/classes.py:1915: FileNotFoundError
```

cmgosnell commented:
current validation failure:

```
============================================================================= short test summary info ==============================================================================
FAILED test/validate/fbp_ferc1_test.py::test_fbp_ferc1_mismatched_fuels[ferc1_annual] - AssertionError: Too many records (8.38%) have mismatched primary fuel types.
FAILED test/validate/fbp_ferc1_test.py::test_vs_bounds[ferc1_annual-oil_cost_per_mmbtu] - ValueError: 10% quantile (3.4921309412014603) is below lower bound (4.0) in validation entitled Oil cost per MMBTU (Tails)
FAILED test/validate/ferc1_test.py::test_no_null_cols_ferc1[ferc1_annual-fbp_ferc1-all] - ValueError: Null columns found in fbp_ferc1: ['other_cost_per_mmbtu']
FAILED test/validate/ferc1_test.py::test_no_null_cols_ferc1[ferc1_annual-plant_in_service_ferc1-all] - KeyError: "['amount_type'] not in index"
FAILED test/validate/ferc1_test.py::test_no_null_cols_ferc1[ferc1_annual-plants_all_ferc1-all] - ValueError: Table plants_small_ferc1 not found
FAILED test/validate/ferc1_test.py::test_no_null_cols_ferc1[ferc1_annual-plants_small_ferc1-all] - ValueError: Table plants_small_ferc1 not found
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-fbp_ferc1-20573] - ValueError: fbp_ferc1: found 25430 rows, expected 20573. Off by 23.609%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-fuel_ferc1-31265] - ValueError: fuel_ferc1: found 48280 rows, expected 31265. Off by 54.422%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-plant_in_service_ferc1-26933] - KeyError: "['amount_type'] not in index"
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-plants_all_ferc1-51456] - ValueError: Table plants_small_ferc1 not found
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-plants_hydro_ferc1-6782] - ValueError: plants_hydro_ferc1: found 6797 rows, expected 6782. Off by 0.221%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-plants_pumped_storage_ferc1-710] - ValueError: plants_pumped_storage_ferc1: found 544 rows, expected 710. Off by -23.380%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-plants_small_ferc1-15584] - ValueError: Table plants_small_ferc1 not found
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-plants_steam_ferc1-28380] - ValueError: plants_steam_ferc1: found 30701 rows, expected 28380. Off by 8.178%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-pu_ferc1-6946] - ValueError: pu_ferc1: found 7423 rows, expected 6946. Off by 6.867%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_minmax_rows[ferc1_annual-purchased_power_ferc1-190228] - ValueError: purchased_power_ferc1: found 197523 rows, expected 190228. Off by 3.835%, allowed margin of 0.000%
FAILED test/validate/ferc1_test.py::test_unique_rows_ferc1[ferc1_annual-plants_hydro_ferc1-unique_subset2] - ValueError: Found 1 dupes of ['report_year', 'utility_id_ferc1', 'plant_name_ferc1', 'capacity_mw'] in dataframe plants_hydro_ferc1
FAILED test/validate/ferc1_test.py::test_unique_rows_ferc1[ferc1_annual-plant_in_service_ferc1-unique_subset4] - KeyError: "['amount_type'] not in index"
FAILED test/validate/fuel_ferc1_test.py::test_vs_bounds[ferc1_annual-gas_cost_per_unit] - ValueError: 5% quantile (1.2239720675660863) is below lower bound (2.0) in validation entitled Cost per unit burned (Gas)
FAILED test/validate/plants_steam_ferc1_test.py::test_vs_bounds[ferc1_annual-capacity_ratios] - ValueError: 5% quantile (0.5694444444444444) is below lower bound (0.64) in validation entitled Capability Ratio (tails)
FAILED test/validate/plants_steam_ferc1_test.py::test_plant_id_clash[ferc1_annual] - AssertionError: Found 34 plant_id_ferc1 values associated with 68 non-unique plant_id_pudl values.
FAILED test/validate/ferc1_test.py::test_record_id_dupes[plants_small_ferc1] - sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) near "plants_small_ferc1": syntax error
FAILED test/validate/ferc1_test.py::test_record_id_dupes[purchased_power_ferc1] - AssertionError: 1 duplicate record_ids found in purchased_power_ferc1: ['purchased_power_326_2021_c001421_cheyenne_light_fuel__power_os'].
```

zaneselvans commented:
The duplicates in the hydro and purchased power tables are coming from the DBF data, not the XBRL data. We've punted on figuring out why they are duplicates, but I think they actually should be addressed (and there are only a tiny handful of them)

cmgosnell commented:
> The duplicates in the hydro and purchased power tables are coming from the DBF data, not the XBRL data. We've punted on figuring out why they are duplicates, but I think they actually should be addressed (and there are only a tiny handful of them)

* the purchased power duplicate is a 2021 (xbrl) dupe and appears to be two legitimate records. We do not necessarily expect the purchase power table to have a unique record_id in the xbrl data so we removed it from this test
* the hydro table was a real dupe from 2019. we've added a targeted removal of one the duplicates which had less data in it.

zaneselvans commented:
@cmgosnell some findings...
* FAILED test/validate/plants_steam_ferc1_test.py::test_vs_bounds[ferc1_annual-capacity_ratios] - ValueError: 5% quantile (0.5694444444444444) is below lower bound (0.64) in validation entitled Capability Ratio (tails)
  * I think this is just due to the fact that coal capacity factors are plummeting. I'd revise the threshold to 50%.
* FAILED test/validate/fbp_ferc1_test.py::test_fbp_ferc1_mismatched_fuels[ferc1_annual] - AssertionError: Too many records (8.38%) have mismatched primary fuel types.
  * The problem here is that a lot of missing values are empty strings, and `"oil" != ""` so the number of mismatches is artificially high.

cmgosnell commented:
and then there were 9:

```
=============================================================================================== short test summary info ================================================================================================
FAILED test/validate/fbp_ferc1_test.py::test_fbp_ferc1_mismatched_fuels[ferc1_annual] - AssertionError: Too many records (8.37%) have mismatched primary fuel types.
FAILED test/validate/fbp_ferc1_test.py::test_fbp_ferc1_mmbtu_cost_correlation[ferc1_annual] - ValueError: oil cost v mmbtu corrcoef is below 0.9: 0.8898862970859709
FAILED test/validate/fbp_ferc1_test.py::test_vs_bounds[ferc1_annual-oil_cost_per_mmbtu] - ValueError: 10% quantile (3.6009842831755097) is below lower bound (4.0) in validation entitled Oil cost per MMBTU (Tails)
FAILED test/validate/ferc1_test.py::test_no_null_cols_ferc1[ferc1_annual-fbp_ferc1-all] - ValueError: Null columns found in fbp_ferc1: ['other_fraction_cost', 'other_fraction_mmbtu', 'other_cost_per_mmbtu']
FAILED test/validate/fuel_ferc1_test.py::test_vs_bounds[ferc1_annual-gas_cost_per_unit] - ValueError: 5% quantile (1.2483937433974757) is below lower bound (2.0) in validation entitled Cost per unit burned (Gas)
FAILED test/validate/plants_steam_ferc1_test.py::test_vs_bounds[ferc1_annual-capacity_ratios] - ValueError: 5% quantile (0.5694444444444444) is below lower bound (0.64) in validation entitled Capability Ratio (tails)
FAILED test/validate/plants_steam_ferc1_test.py::test_plant_id_clash[ferc1_annual] - AssertionError: Found 34 plant_id_ferc1 values associated with 68 non-unique plant_id_pudl values.
============================================================== 7 failed, 250 passed, 124 skipped, 7 xfailed, 1 xpassed, 2 warnings in 1825.64s (0:30:25) ===============================================================
[1]    90996 exit 1     pytest test/validate --live-dbs
```

zaneselvans commented:
I think the unexpectedly large low-end tail in natural gas prices may be due to some unfixed unit errors. There's a cluster of fuel prices reported between 0-0.5 USD/Mcf that looks suspiciously similar to the main distribution between 0-5 USD/Mcf. So maybe it was reported in USD / hundred cubic feet, rather than USD / thousand cubic feet?

![image](https://user-images.githubusercontent.com/596279/202544371-661602f3-2331-4cf7-aa28-6ea8213ea4f0.png)
![image](https://user-images.githubusercontent.com/596279/202544435-d2802686-bb35-49f2-803e-61b9f0b4da52.png)

cmgosnell commented:
and then there were 4:
```
FAILED test/validate/fbp_ferc1_test.py::test_fbp_ferc1_mmbtu_cost_correlation[ferc1_annual] - ValueError: nuclear cost v mmbtu corrcoef is below 0.9: 0.09498177057189967
FAILED test/validate/fbp_ferc1_test.py::test_vs_bounds[ferc1_annual-oil_cost_per_mmbtu] - ValueError: 10% quantile (3.6009842831755097) is below lower bound (4.0) in validation entitled Oil cost per MMBTU (Tails)
FAILED test/validate/fuel_ferc1_test.py::test_vs_bounds[ferc1_annual-gas_cost_per_unit] - ValueError: 5% quantile (1.2483937433974757) is below lower bound (2.0) in validation entitled Cost per unit burned (Gas)
FAILED test/validate/plants_steam_ferc1_test.py::test_plant_id_clash[ferc1_annual] - AssertionError: Found 34 plant_id_ferc1 values associated with 68 non-unique plant_id_pudl values.
```

zaneselvans commented:
I think the oil price distribution actually just includes more lower priced records now. Probably as a result of all the fuel type / units cleaning.  There's a little spike at ~0 that's probably bad data, but removing it doesn't make much of a difference in where the 10% cutoff is. So I would say drop the 10% cutoff from 4.0 USD/MMBtu to 3.0 USD/MMBtu on this one:
```
 FAILED test/validate/fbp_ferc1_test.py::test_vs_bounds[ferc1_annual-oil_cost_per_mmbtu] - ValueError: 10% quantile (3.4921309412014603) is below lower bound (4.0) in validation entitled Oil cost per MMBTU (Tails)
```
![image](https://user-images.githubusercontent.com/596279/202551154-f16309cd-12cd-4e45-945b-587aed085f5f.png)

zaneselvans commented:
On the cost correlations, I think you can just remove `nuclear` and `other` from the fuel types that are checked. Nuclear fuel reporting is a mess in FERC 1. There's really no good data to use.  And we eliminated all of the `other` fuel records.

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -1217,6 +1217,48 @@ class PlantsHydroFerc1TableTransformer(Ferc1AbstractTableTransformer):
+        eachother, except one have nulls in the capex columns. Surgically remove the
+        record with the nulls.
+        """
+        if 2019 in df.report_year.unique():
```
i made this run only if 2019 is being processed, but I could change the assertion below to only yell if `len(possible_dupes) != 2` and `2019 in df.report_year.unique()`

zaneselvans replied:
I think only having the assertion in the conditional is better -- we want these kinds of spot fixes to run and work regardless of what data is being processed, and having the whole thing enclosed in a conditional suggests that it will *not* work if 2019 isn't in the data.

cmgosnell commented on test/conftest.py:
```diff
@@ -292,13 +292,16 @@ def pudl_settings_dict(request, live_dbs, tmpdir_factory):  # noqa: C901
-            "ferc1_xbrl_taxonomy_metadata"
-        ]
-        pudl_settings["censusdp1tract_db"] = pudl_defaults["censusdp1tract_db"]
+        # everything with the following suffixes should use the defaults as opposed to
+        # the generated settings. We should overhaul using temp_dir's for pudl_out at
+        # all while using `live_dbs` and perhaps change the name of `live_dbs` bc it
+        # now encompasses more than just dbs
+        overwrite_suffixes = ("_db", "_datapackage", "_xbrl_taxonomy_metadata")
+        pudl_settings = {
+            k: pudl_defaults[k] if k.endswith(overwrite_suffixes) else v
+            for (k, v) in pudl_settings.items()
+        }
+
```
decided to go with this dict comprehension here instead of overhauling how we set `pudl_out` / what `live_dbs` means because TIME. but we should definitely do that overhaul

cmgosnell commented on test/validate/ferc1_test.py:
```diff
@@ -65,16 +63,16 @@ def test_no_null_cols_ferc1(pudl_out_ferc1, live_dbs, cols, df_name):
+        ("plant_in_service_ferc1", 311_794),
+        ("plants_all_ferc1", 54_276),
+        ("plants_hydro_ferc1", 6_796),
+        ("plants_pumped_storage_ferc1", 544),
```
this is the only one that actually went down. i believe this is because we are actually dropping rows now!

zaneselvans commented on test/validate/fuel_ferc1_test.py:
```diff
@@ -37,7 +37,11 @@ def test_fuel_ferc1_trivial(pudl_out_ferc1):
             marks=pytest.mark.xfail(reason="FERC 1 fuel unit errors?"),
         ),
-        pytest.param(pv.fuel_ferc1_gas_cost_per_unit_bounds, id="gas_cost_per_unit"),
+        pytest.param(
```
Can you put a comment in here with the full URL for the issue? Then it's clickable on GitHub.

zaneselvans reviewed (changes requested):
Only the tiniest of changes requested!
